### PR TITLE
Added option to enable cluster proportional autoscaling of core dns.

### LINKED ISSUE #4084: Evaluate dns-autoscaler as replacement of HPA for CoreDNS
**How to categorize this issue?**

/area auto-scaling networking
/kind enhancement
/priority 3

**What would you like to be added**:
Evaluate dns-autoscaler as replacement of HPA for CoreDNS.

Currently simple [horizontal](https://github.com/gardener/gardener/blob/v1.23.0/charts/shoot-core/components/charts/coredns/templates/coredns-hpa.yaml) pod autoscaler that scales based on the CPU usage. The community provides a [guideline](https://kubernetes.io/docs/tasks/administer-cluster/dns-horizontal-autoscaling/) how dns-autoscaler to be used for the CoreDNS. This solution also respect the size of the cluster and can help in large clusters.

**Why is this needed**:

### PR DESCRIPTION
**Pull request by ScheererJ**
For large clusters, it may be beneficial to scale core dns according to the cluster
size. This change allows to switch between the standard horizontal autoscaling of
core dns using cpu as metric or the cluster proportional autoscaler.

**How to categorize this PR?**

/area networking
/kind enhancement

**What this PR does / why we need it**:
Added option to enable cluster proportional autoscaling of core dns.

For large clusters, it may be beneficial to scale core dns according to the cluster
size. This change allows to switch between the standard horizontal autoscaling of
core dns using cpu as metric or the cluster proportional autoscaler.

**Which issue(s) this PR fixes**:
Fixes #4084 

**Special notes for your reviewer**:

**Release note**:

```other operator
Allows cluster owner to switch between horizontal and cluster-proportional autoscaling of coredns
```

### CODE REVIEW

**Comment by ScheererJ** [comment]
Tests are currently missing.

File: docs/usage/dns-autoscaling.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# DNS Autoscaling
```
Comment (rfranzke) [inline docs/usage/dns-autoscaling.md:1]: Please add this document to the index (README.md)

File: docs/usage/dns-autoscaling.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+This is a short guide describing different options how to automatically scale DNS in the shoot cluster.
```
Comment (rfranzke) [inline docs/usage/dns-autoscaling.md]:
```suggestion
This is a short guide describing different options how to automatically scale CoreDNS in the shoot cluster.
```

File: docs/usage/dns-autoscaling.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+```
+replicas = max( ceil( cores × 1 / coresPerReplica ) , ceil( nodes × 1 / nodesPerReplica ) )
+```
+
+Depending on your needs, you can adjust `coresPerReplica` or `nodesPerReplica`, but it is also possible to override `min` if required.
+
+## Trade-offs of horizontal and cluster-proportional DNS Autoscaling
+
+The horizontal autoscaling of CoreDNS as implemented by Gardener is fully managed, i.e. you do not need to perform any configuration changes. It scales according to the cpu usage of CoreDNS replicas meaning that it will create new replicas if the existing ones are under heavy load. However, it is limited to scale between 2 and 5 instances.
```
Comment (rfranzke) [inline docs/usage/dns-autoscaling.md]:
```suggestion
The horizontal autoscaling of CoreDNS as implemented by Gardener is fully managed, i.e. you do not need to perform any configuration changes. It scales according to the CPU usage of CoreDNS replicas meaning that it will create new replicas if the existing ones are under heavy load. However, it is limited to scale between 2 and 5 instances.
```

Comment (rfranzke) [reply docs/usage/dns-autoscaling.md]: As a reader, I would wonder why it is limited to 5 instances while I can tweak the settings for `cluster-proportional` scaling myself. Why can't I tweak this maximum number?

File: docs/usage/dns-autoscaling.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+The cluster-proportional autoscaling of CoreDNS as implemented by Gardener is fully managed, but allows more configuration options to adjust the default settings to your individual needs. It scales according to the cluster size, i.e. if your cluster grows in terms of cores/nodes so will the amount of CoreDNS replicas. However, it does not take the actual workload, e.g. cpu consumption, into account.
```
Comment (rfranzke) [inline docs/usage/dns-autoscaling.md]:
```suggestion
The cluster-proportional autoscaling of CoreDNS as implemented by Gardener is fully managed, but allows more configuration options to adjust the default settings to your individual needs. It scales according to the cluster size, i.e. if your cluster grows in terms of cores/nodes so will the amount of CoreDNS replicas. However, it does not take the actual workload, e.g. CPU consumption, into account.
```

File: docs/usage/dns-autoscaling.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+Experience shows that the horizontal autoscaling of CoreDNS works for a variety of workloads. It does reach its limits if a cluster has a high amount of DNS requests, though. The cluster-proportional autoscaling approach allows to fine-tune the amount of CoreDNS replicas. It helps to scale in clusters of changing size. However, please keep in mind that you need to cater for the maximum amount of DNS requests as the replicas will not be adapted according to the workload, but only according to the cluster size (cores/nodes).
+
+## Trade-offs in conjunction with NodeLocalDNS
```
Comment (rfranzke) [inline docs/usage/dns-autoscaling.md:59]: Above you write

> Please note that there is also the option to use a [node-local DNS cache](node-local-dns.md), which helps mitigate potential DNS bottlenecks, but is not discussed in detail in this guide.

but now you start discussing it :) Maybe not in the detail you had in mind when writing the above, however, I would adapt above sentence to not confuse readers.

File: pkg/apis/core/types_shoot.go [hunk 6]
```diff
@@ -981,6 +983,39 @@ var (
...
 )
 
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// System components relevant types                                                             //
+//////////////////////////////////////////////////////////////////////////////////////////////////
+
+// SystemComponents contains the settings of system components in the control or data plane of the Shoot cluster.
+type SystemComponents struct {
+	// CoreDNS contains the settings of the Core DNS components running in the data plane of the Shoot cluster.
+	CoreDNS *CoreDNS `json:"coreDNS" protobuf:"bytes,1,opt,name=coreDNS"`
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]: Drop all `json` tags etc from internal APIs

File: pkg/apis/core/v1alpha1/defaults.go [hunk 7]
```diff
@@ -287,6 +287,14 @@ func SetDefaults_Shoot(obj *Shoot) {
 
 		obj.Spec.Provider.Workers[i].CRI = &CRI{Name: CRINameContainerD}
 	}
+
+	if obj.Spec.SystemComponents == nil {
+		obj.Spec.SystemComponents = &SystemComponents{}
+	}
+	if obj.Spec.SystemComponents.CoreDNS == nil {
+		obj.Spec.SystemComponents.CoreDNS = &CoreDNS{}
+		obj.Spec.SystemComponents.CoreDNS.Autoscaling.Mode = CoreDNSAutoscalingModeHorizontal
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/defaults.go:292]: If I set `.spec.systemComponents.coreDNS={}` in my shoot then the autoscaling mode will not be defaulted.

File: pkg/apis/core/v1beta1/defaults.go [hunk 8]
```diff
@@ -287,6 +287,14 @@ func SetDefaults_Shoot(obj *Shoot) {
 
 		obj.Spec.Provider.Workers[i].CRI = &CRI{Name: CRINameContainerD}
 	}
+
+	if obj.Spec.SystemComponents == nil {
+		obj.Spec.SystemComponents = &SystemComponents{}
+	}
+	if obj.Spec.SystemComponents.CoreDNS == nil {
+		obj.Spec.SystemComponents.CoreDNS = &CoreDNS{}
+		obj.Spec.SystemComponents.CoreDNS.Autoscaling.Mode = CoreDNSAutoscalingModeHorizontal
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/defaults.go:292]: same here

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 9]
```diff
@@ -120,6 +120,9 @@ type ShootSpec struct {
 	// ExposureClassName is the optional name of an exposure class to apply a control plane endpoint exposure strategy.
 	// +optional
 	ExposureClassName *string `json:"exposureClassName,omitempty" protobuf:"bytes,18,opt,name=exposureClassName"`
+	// SystemComponents contains the settings of system components in the control or data plane of the Shoot cluster.
+	// +optional
+	SystemComponents *SystemComponents `json:"systemComponents" protobuf:"bytes,19,opt,name=systemComponents"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go:129]: Please adapt the `Shoot` example file in the `examples` folder with this new field

File: pkg/operation/botanist/component/coredns/coredns.go [hunk 10]
```diff
@@ -526,6 +533,137 @@ import custom/*.server
 				},
 			},
 		}
+
+		clusterProportionalDNSAutoscalerServiceAccount = &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "kube-dns-autoscaler",
+				Namespace: metav1.NamespaceSystem,
+			},
```
Comment (rfranzke) [inline pkg/operation/botanist/component/coredns/coredns.go:552]: Set `automountServiceAccountToken=false`

File: pkg/operation/botanist/coredns.go [hunk 11]
```diff
@@ -41,17 +41,27 @@ func (b *Botanist) DefaultCoreDNS() (coredns.Interface, error) {
...
+		PodNetworkCIDR:                    b.Shoot.Networks.Pods.String(),
+		NodeNetworkCIDR:                   b.Shoot.GetInfo().Spec.Networking.Nodes,
+		UseClusterProportionalAutoscaling: false,
 	}
 
 	if b.APIServerSNIEnabled() {
 		values.APIServerHost = pointer.String(b.outOfClusterAPIServerFQDN())
 	}
 
+	if b.Shoot.GetInfo().Spec.SystemComponents != nil && b.Shoot.GetInfo().Spec.SystemComponents.CoreDNS != nil && b.Shoot.GetInfo().Spec.SystemComponents.CoreDNS.Autoscaling.Mode == gardencorev1beta1.CoreDNSAutoscalingModeClusterProportional {
```
Comment (rfranzke) [inline pkg/operation/botanist/coredns.go]: Nit: You could introduce a helper function in `pkg/apis/core/v1beta1/helper` to retrieve the CoreDNS autoscaling mode when providing the `*SystemComponents`.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: docs/usage/dns-autoscaling.md (hunk 3 above)
Comment (ScheererJ) [reply docs/usage/dns-autoscaling.md]: The existing autoscaling solution simply does not allow users to configure the maximum amount of instances. There is no technical reason (I am aware of) for this though.
Do you suggest to make the existing solution more configurable as part of this change request or would you rather like to adapt the documentation?

Comment (rfranzke) [reply docs/usage/dns-autoscaling.md]: > would you rather like to adapt the documentation?

This. While reading it this questions arose in my mind, so maybe we can answer it upfront if others are wondering similarly.

Comment (ScheererJ) [reply docs/usage/dns-autoscaling.md]: Ok, I will try, but frankly I do not understand why it is not configurable.

File: pkg/operation/botanist/component/coredns/coredns.go (hunk 10 above)
Comment (ScheererJ) [reply pkg/operation/botanist/component/coredns/coredns.go:552]: Why should the service account token not by automounted? The autoscaler needs access to the `kube-apiserver` to get the cluster size in terms of nodes and scale CoreDNS accordingly.
In other words: It does not seem to work if I set `automountServiceAccountToken=false`, but crashes instead:
```
E0125 10:26:47.318985       1 autoscaler.go:52] open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
```

File: docs/usage/dns-autoscaling.md (hunk 1 above)
Comment (ScheererJ) [reply docs/usage/dns-autoscaling.md:1]: Done with newest commit.

File: docs/usage/dns-autoscaling.md (hunk 2 above)
Comment (ScheererJ) [reply docs/usage/dns-autoscaling.md]: Updated accordingly.

File: docs/usage/dns-autoscaling.md (hunk 3 above)
Comment (ScheererJ) [reply docs/usage/dns-autoscaling.md]: Fixed the casing of `CPU` and adapted the wording of the scale limits of the horizontal autoscaling solution.
Feedback is welcome.

File: docs/usage/dns-autoscaling.md (hunk 4 above)
Comment (ScheererJ) [reply docs/usage/dns-autoscaling.md]: Fixed.

File: docs/usage/dns-autoscaling.md (hunk 5 above)
Comment (ScheererJ) [reply docs/usage/dns-autoscaling.md:59]: Rephrased to be hopefully less confusing.

File: pkg/apis/core/v1alpha1/defaults.go (hunk 7 above)
Comment (ScheererJ) [reply pkg/apis/core/v1alpha1/defaults.go:292]: Adapted. I hope this is the right way to do it.

File: pkg/apis/core/v1beta1/defaults.go (hunk 8 above)
Comment (ScheererJ) [reply pkg/apis/core/v1beta1/defaults.go:292]: Same as above.

File: pkg/apis/core/types_shoot.go (hunk 6 above)
Comment (ScheererJ) [reply pkg/apis/core/types_shoot.go]: Removed the `json` tags from the internal APIs.

File: pkg/apis/core/v1beta1/types_shoot.go (hunk 9 above)
Comment (ScheererJ) [reply pkg/apis/core/v1beta1/types_shoot.go:129]: `Shoot` example adapted with `systemComponents` section.

File: pkg/operation/botanist/coredns.go (hunk 11 above)
Comment (ScheererJ) [reply pkg/operation/botanist/coredns.go]: Extracted a helper function as suggested.

**Comment by ScheererJ** [comment]
Tests added.

**Comment by rfranzke** [comment]
Can you check the failing `verify` step?

File: example/90-shoot.yaml [hunk 12]
```diff
@@ -344,3 +344,7 @@ spec:
 #     kind: Secret
 #     name: my-foobar-secret
 # exposureClassName: <exposure-class-name>
+# systemComponents:
+#   CoreDNS:
+#     Autoscaling:
+#       Mode: horizontal # {horizontal,cluster-proportional}
```
Comment (rfranzke) [inline example/90-shoot.yaml]:
```suggestion
#   coreDNS:
#     autoscaling:
#       mode: horizontal # {horizontal,cluster-proportional}
```

File: pkg/admissioncontroller/webhooks/admission/resourcesize/admission_test.go [hunk 13]
```diff
@@ -60,9 +60,9 @@ var _ = Describe("handler", func() {
 		projectsSizeLimit, _ = resource.ParseQuantity("0M")
 		secretSizeLimit, _   = resource.ParseQuantity("1Mi")
 		// size of shoot w/ namespace, name, w/o spec
-		shootsv1beta1SizeLimit, _ = resource.ParseQuantity("405")
+		shootsv1beta1SizeLimit, _ = resource.ParseQuantity("429")
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhooks/admission/resourcesize/admission_test.go:63]: Out of curiosity: Why was this changed?

File: pkg/apis/core/v1beta1/helper/helper.go [hunk 14]
```diff
@@ -1473,3 +1473,16 @@ func AddTypeToSecretBinding(secretBinding *gardencorev1beta1.SecretBinding, prov
 	}
 	secretBinding.Provider.Type = strings.Join(types, ",")
 }
+
+// IsCoreDNSClusterProportionalAutoscalingEnabled indicates whether cluster-proportional CoreDNS autoscaling is enabled or not.
+func IsCoreDNSClusterProportionalAutoscalingEnabled(systemComponents *gardencorev1beta1.SystemComponents) bool {
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/helper/helper.go]: Does it make sense to make this function more general so that it can be used for other autoscaling modes as well?
```suggestion
func IsCoreDNSAutoscalingUsed(systemComponents *gardencorev1beta1.SystemComponents, autoscalingMode gardencorev1beta1.CoreDNSAutoscalingMode) bool {
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 15]
```diff
@@ -1237,6 +1240,40 @@ var (
...
+type SystemComponents struct {
+	// CoreDNS contains the settings of the Core DNS components running in the data plane of the Shoot cluster.
+	// +optional
+	CoreDNS *CoreDNS `json:"coreDNS" protobuf:"bytes,1,opt,name=coreDNS"`
+}
+
+// CoreDNS contains the settings of the Core DNS components running in the data plane of the Shoot cluster.
+type CoreDNS struct {
+	// Autoscaling contains the settings related to autoscaling of the Core DNS components running in the data plane of the Shoot cluster.
+	Autoscaling CoreDNSAutoscaling `json:"autoscaling" protobuf:"bytes,1,opt,name=autoscaling"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]: Shouldn't this be a pointer (= optional) as well? 😕

File: pkg/operation/botanist/component/coredns/coredns.go [hunk 16]
```diff
@@ -90,6 +93,10 @@ type Values struct {
 	PodNetworkCIDR string
 	// NodeNetworkCIDR is the CIDR of the node network.
 	NodeNetworkCIDR *string
+	// UseClusterProportionalAutoscaling indicates whether cluster proportional autoscaling is enabled.
+	UseClusterProportionalAutoscaling bool
```
Comment (rfranzke) [inline pkg/operation/botanist/component/coredns/coredns.go]: Maybe just pass the autoscaling mode (instead of hard-coding one of many possible modes)?

File: pkg/operation/botanist/component/coredns/coredns.go (hunk 10 above)
Comment (rfranzke) [reply pkg/operation/botanist/component/coredns/coredns.go:552]: Setting it to `false` should trigger the projected token mount (https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#auto-mounting-projected-serviceaccount-tokens) and static token invalidation). Did you rebase to `master` when testing this?

File: pkg/admissioncontroller/webhooks/admission/resourcesize/admission_test.go (hunk 13 above)
Comment (ScheererJ) [reply pkg/admissioncontroller/webhooks/admission/resourcesize/admission_test.go:63]: Well, the test "should pass because size is in range for v1beta1 shoot" checks the size of a more or less empty shoot specification. However, with this change `,"systemComponents":null` is added. Therefore, the resourceSize is 24 bytes bigger.
(I also increased the v1alpha1 size for consistency. No test actually checks it.)

Comment (rfranzke) [reply pkg/admissioncontroller/webhooks/admission/resourcesize/admission_test.go:63]: :-) Thanks!

File: example/90-shoot.yaml (hunk 12 above)
Comment (ScheererJ) [reply example/90-shoot.yaml]: Good catch.
Will be fixed in the next commit.

File: pkg/apis/core/v1beta1/helper/helper.go (hunk 14 above)
Comment (ScheererJ) [reply pkg/apis/core/v1beta1/helper/helper.go]: Will be adapted in next commit.

File: pkg/apis/core/v1beta1/types_shoot.go (hunk 15 above)
Comment (ScheererJ) [reply pkg/apis/core/v1beta1/types_shoot.go]: Ok. Done with next commit.

File: pkg/operation/botanist/component/coredns/coredns.go (hunk 16 above)
Comment (ScheererJ) [reply pkg/operation/botanist/component/coredns/coredns.go]: Adapted with next commit.

File: pkg/operation/botanist/component/coredns/coredns.go (hunk 10 above)
Comment (ScheererJ) [reply pkg/operation/botanist/component/coredns/coredns.go:552]: I just rebased and set everything up according to the link you provided, i.e. added `automountServiceAccountToken=false` to the deployment as well as the service account, but it does not work for me :-( .

Comment (ScheererJ) [reply pkg/operation/botanist/component/coredns/coredns.go:552]: Updated with newest commit.

**Comment by ScheererJ** [comment]
Renamed `kube-dns-autoscaler` to `core-dns-autoscaler` to avoid confusion.

File: docs/usage/dns-autoscaling.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  systemComponents:
+    coreDNS:
+      autoscaling:
+        mode: cluster-proportional
+...
+```
+
+To switch back to horizontal DNS autoscaling you can set the `spec.systemComponents.coreDNS.autoscaling.mode` to `horizontal` (or remove the `coreDNS` section).
+
+Once the cluster-proportional autoscaling of CoreDNS has been enabled and the Shoot cluster has been reconciled afterwards, a ConfigMap called `core-dns-autoscaler` will be created in the `kube-system` namespace with the default settings. The content will be similar to the following:
```
Comment (rfranzke) [inline docs/usage/dns-autoscaling.md]: Nit: Remove the `-` in `core-dns` since everything else is called `coredns*`.

Comment (ScheererJ) [reply docs/usage/dns-autoscaling.md]: Renamed with latest commit.

File: pkg/apis/core/v1beta1/defaults.go [hunk 18]
```diff
@@ -278,6 +278,19 @@ func SetDefaults_Shoot(obj *Shoot) {
...
+	}
+	if obj.Spec.SystemComponents.CoreDNS == nil {
+		obj.Spec.SystemComponents.CoreDNS = &CoreDNS{}
+	}
+	if obj.Spec.SystemComponents.CoreDNS.Autoscaling == nil {
+		obj.Spec.SystemComponents.CoreDNS.Autoscaling = &CoreDNSAutoscaling{}
+	}
+	if obj.Spec.SystemComponents.CoreDNS.Autoscaling.Mode != CoreDNSAutoscalingModeHorizontal && obj.Spec.SystemComponents.CoreDNS.Autoscaling.Mode != CoreDNSAutoscalingModeClusterProportional {
+		obj.Spec.SystemComponents.CoreDNS.Autoscaling.Mode = CoreDNSAutoscalingModeHorizontal
+	}
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/defaults.go:293]: Can you please add a unit test for this defaulting (same in `v1alpha1`)?

File: pkg/apis/core/v1beta1/helper/helper.go [hunk 19]
```diff
@@ -1472,3 +1472,21 @@ func AddTypeToSecretBinding(secretBinding *gardencorev1beta1.SecretBinding, prov
 	}
 	secretBinding.Provider.Type = strings.Join(types, ",")
 }
+
+// IsCoreDNSAutoscalingUsed indicates whether the specified autoscaling mode of CoreDNS is enabled or not.
+func IsCoreDNSAutoscalingUsed(systemComponents *gardencorev1beta1.SystemComponents, autoscalingMode gardencorev1beta1.CoreDNSAutoscalingMode) bool {
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/helper/helper.go]:
```suggestion
func IsCoreDNSAutoscalingModeUsed(systemComponents *gardencorev1beta1.SystemComponents, autoscalingMode gardencorev1beta1.CoreDNSAutoscalingMode) bool {
```

File: pkg/apis/core/validation/shoot.go [hunk 20]
```diff
@@ -1609,3 +1614,31 @@ func ValidateContainerRuntimes(containerRuntime []core.ContainerRuntime, fldPath
 
 	return allErrs
 }
+
+// validateSystemComponents validates the given system components.
+func validateSystemComponents(systemComponents *core.SystemComponents, fldPath *field.Path) field.ErrorList {
```
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go]: unit test?

File: pkg/operation/botanist/component/coredns/coredns.go [hunk 21]
```diff
@@ -528,6 +536,138 @@ import custom/*.server
...
+				Name:     clusterProportionalDNSAutoscalerClusterRole.Name,
+			},
+			Subjects: []rbacv1.Subject{{
+				Kind:      rbacv1.ServiceAccountKind,
+				Name:      clusterProportionalDNSAutoscalerServiceAccount.Name,
+				Namespace: clusterProportionalDNSAutoscalerServiceAccount.Namespace,
+			}},
+		}
+
+		clusterProportionalDNSAutoscalerDeployment = &appsv1.Deployment{
```
Comment (rfranzke) [inline pkg/operation/botanist/component/coredns/coredns.go:602]: Can you add a `VPA` for this deployment in case the shoot enables VPA?

File: pkg/operation/botanist/component/coredns/coredns.go [hunk 22]
```diff
@@ -528,6 +536,138 @@ import custom/*.server
...
+						Containers: []corev1.Container{{
+							Name:            "autoscaler",
+							Image:           c.values.ClusterProportionalAutoscalerImage,
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Command: []string{
+								"/cluster-proportional-autoscaler",
+								"--namespace=" + metav1.NamespaceSystem,
+								"--configmap=coredns-autoscaler",
+								"--target=deployment/" + deployment.Name,
+								"--default-params={\"linear\":{\"coresPerReplica\":256,\"nodesPerReplica\":16,\"min\":2,\"preventSinglePointFailure\":true,\"includeUnschedulableNodes\":true}}",
```
Comment (rfranzke) [inline pkg/operation/botanist/component/coredns/coredns.go]: Nit: Use \` to avoid escaping `"`

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/apis/core/v1beta1/defaults.go (hunk 18 above)
Comment (ScheererJ) [reply pkg/apis/core/v1beta1/defaults.go:293]: Added test with newest commit for v1beta1 as well as for v1alpha1.

File: pkg/apis/core/v1beta1/helper/helper.go (hunk 19 above)
Comment (ScheererJ) [reply pkg/apis/core/v1beta1/helper/helper.go]: Renamed accordingly with latest commit.

File: pkg/apis/core/validation/shoot.go (hunk 20 above)
Comment (ScheererJ) [reply pkg/apis/core/validation/shoot.go]: Unit test added with latest commit.

File: pkg/operation/botanist/component/coredns/coredns.go (hunk 22 above)
Comment (ScheererJ) [reply pkg/operation/botanist/component/coredns/coredns.go]: Changed to using ` for the line including quotes with latest commit.

File: pkg/operation/botanist/component/coredns/coredns.go (hunk 21 above)
Comment (ScheererJ) [reply pkg/operation/botanist/component/coredns/coredns.go:602]: Added vertical pod autoscaler in latest commit.
However, it feels somewhat strange to autoscale the autoscaler. It also makes the values more confusing.

**Review by rfranzke — APPROVED** [review]
/lgtm
